prog/dftb+/prg_dftb/mainio.F90

```diff
@@ -82,524 +96,1459 @@ module mainio
+    !> Index array for sparse matrices
+    integer, intent(in) :: iSparseStart(:,:)
 
-    !> If present, Hamiltonian(s) are fetched from this storage into HSqrReal, instead of using
-    !> whatever is already there
-    type(OFifoRealR2), intent(inout), optional :: storeEigvecs(:)
+    !> Mapping of atoms into the central cell.
+    integer, intent(in) :: img2CentCell(:)
 
-    integer, allocatable :: iOrbs(:)
-    integer :: iSpin, iSpin2, iLev, ii, nReg, dummy
-    integer :: valshape(1)
-    real(dp) :: qState
-    real(dp), allocatable :: rVecTemp(:)
-    character(lc) :: tmpStr
+    !> Species of each atom
+    integer, intent(in) :: species(:)
+
+    !> Name of each species
+    character(*), intent(in) :: speciesName(:)
+
+    !> optional alternative file prefix, to appear as "fileName".bin
+    character(len=*), intent(in), optional :: fileName
+
+    type(linecomm) :: collector
+    real(dp), allocatable :: localEigvec(:), localFrac(:)
+    real(dp), allocatable :: globalS(:,:), globalFrac(:,:)
+    integer :: nOrb, nAtom
+    integer :: iKS, iS, iGroup, iEig
+
+    nOrb = denseDesc%fullSize
+    nAtom = size(nNeighbor)
+    allocate(globalS(size(eigvecs, dim=1), size(eigvecs, dim=2)))
+    allocate(globalFrac(size(eigvecs, dim=1), size(eigvecs, dim=2)))
+    if (env%mpi%tGroupMaster) then
+      allocate(localEigvec(nOrb))
+      allocate(localFrac(nOrb))
+    end if
+
+    call collector%init(env%blacs%orbitalGrid, denseDesc%blacsOrbSqr, "c")
+    if (env%mpi%tGlobalMaster) then
+      call prepareEigvecFileTxt(fd, .false., fileName)
+    end if
+
+    masterSlave: if (env%mpi%tGlobalMaster) then
+      do iKS = 1, parallelKS%maxGroupKS
+        group: do iGroup = 0, env%mpi%nGroup - 1
+          if (iKS > parallelKS%nGroupKS(iGroup)) then
+            cycle group
+          end if
+          iS = parallelKS%groupKS(2, iKS, iGroup)
+          if (iGroup == 0) then
+            call unpackHSRealBlacs(env%blacs, over, iNeighbor, nNeighbor, iSparseStart,&
+                & img2CentCell, denseDesc, globalS)
+            call pblasfx_psymm(globalS, denseDesc%blacsOrbSqr, eigvecs(:,:,iKS),&
+                & denseDesc%blacsOrbSqr, globalFrac, denseDesc%blacsOrbSqr)
+            globalFrac(:,:) = globalFrac * eigvecs(:,:,iKS)
+          end if
+          do iEig = 1, nOrb
+            if (iGroup == 0) then
+              call collector%getline_master(env%blacs%orbitalGrid, iEig, eigvecs(:,:,iKS),&
+                  & localEigvec)
+              call collector%getline_master(env%blacs%orbitalGrid, iEig, globalFrac, localFrac)
+            else
+              call mpifx_recv(env%mpi%interGroupComm, localEigvec, iGroup)
+              call mpifx_recv(env%mpi%interGroupComm, localFrac, iGroup)
+            end if
+            call writeSingleRealEigvecTxt(fd, localEigvec, localFrac, iS, iEig, orb, species,&
+                & speciesName, nAtom)
+          end do
+        end do group
+      end do
+    else
```

Comment:
Comment on which case of the if this branch is for

Reply:
Done.